core/src/main/java/io/grpc/Context.java

```diff
@@ -0,0 +1,653 @@
+      if (key.equals(keyValueEntries[i][0])) {
+        return keyValueEntries[i][1];
+      }
+    }
+    if (parent == null) {
+      return null;
+    }
+    return parent.lookup(key);
+  }
+
+  /**
+   * A context which inherits cancellation from its parent but which can also be independently
+   * cancelled and which will propagate cancellation to it's descendants.
+   */
+  public static final class CancellableContext extends Context {
+
+    /**
+     * Used with cancelledUpdater. 0->not cancelled, 1->cancelled.
+     */
+    @SuppressWarnings("unused")
+    private boolean cancelled;
+    private Throwable cause;
+    private final Context dummy;
+    private ScheduledFuture<?> scheduledFuture;
+
+    /**
+     * Create a cancellable context that does not have a deadline.
+     */
+    private CancellableContext(Context parent) {
+      super(parent, EMPTY_ENTRIES);
+      // Create a dummy that inherits from this to attach and detach so that you cannot retrieve a
+      // cancellable context from Context.current()
+      dummy = new Context(this, EMPTY_ENTRIES);
+    }
+
+    /**
+     * Create a cancellable context that has a deadline.
+     */
+    private CancellableContext(Context parent, long delayNanos) {
+      super(parent, EMPTY_ENTRIES);
+      // Create a dummy that inherits from this to attach and detach so that you cannot retrieve a
+      // cancellable context from Context.current()
+      dummy = new Context(this, EMPTY_ENTRIES);
+      scheduledFuture = SCHEDULER.create().schedule(new Runnable() {
+        @Override
+        public void run() {
+          cancel(new TimeoutException("context timed out"));
+        }
+      }, delayNanos, TimeUnit.NANOSECONDS);
+    }
+
+
+    @Override
+    public void attach() {
+      dummy.attach();
+    }
+
+    @Override
+    public void detach() {
+      dummy.detach();
+    }
+
+    @Override
+    public boolean isCurrent() {
+      return dummy.isCurrent();
+    }
+
+    /**
+     * Attach this context to the thread and return a {@link java.lang.AutoCloseable} that can be
+     * used with try-with-resource statements to properly {@link #detach} and {@link #cancel}
+     * the context on completion.
+     *
+     * @return a {@link java.io.Closeable} which can be used with try-with-resource blocks.
+     */
+    public Closeable attachAsCloseable() {
+      attach();
+      return new Closeable() {
+        @Override
+        public void close() throws IOException {
+          detachAndCancel(null);
+        }
+      };
+    }
+
+    /**
+     * Cancel this context and optionally provide a cause for the cancellation, all listeners
+     * are notified immediately on the calling thread.
+     *
+     * @return {@code true} if this context cancelled the context and notified listeners,
+     *    {@code false} if the context was already cancelled.
+     */
+    public boolean cancel(@Nullable Throwable cause) {
+      synchronized (this) {
+        if (!cancelled) {
```

Comment:
It's hard to tell, but should we call isCancelled() here? If this call to cancel was triggered by a listener being called, it seems it would probably be good if we set the cause to be that of the parent.